src/WallToolPaths.cpp

```diff
@@ -82,24 +88,116 @@ const VariableWidthPaths& WallToolPaths::getToolPaths()
+    for(VariableWidthLines& path : contour_paths)
+    {
+        for(ExtrusionLine& line : path)
+        {
+            line_endpoints.insert(&line);
+        }
+    }
+    //Then go through all lines and construct chains of polylines if the endpoints are nearby.
+    std::unordered_set<ExtrusionLine*> processed_lines; //Track which lines were already processed to not process them twice.
+    for(VariableWidthLines& path : contour_paths)
+    {
+        for(ExtrusionLine& line : path)
+        {
+            if(processed_lines.find(&line) != processed_lines.end()) //We already added this line before.
+            {
+                continue;
+            }
+            inner_contour.emplace_back();
+            ExtrusionLine* nearest = &line;
+            while(nearest)
+            {
+                if(processed_lines.find(nearest) != processed_lines.end())
+                {
+                    break; //Looping. This contour is already processed.
+                }
+
+                for(const ExtrusionJunction& junction : nearest->junctions)
+                {
+                    inner_contour.back().add(junction.p);
+                }
+                processed_lines.insert(nearest);
+
+                //Find any nearby lines to attach.
+                const Point current_position = nearest->junctions.back().p;
+                const std::vector<ExtrusionLine*> nearby = line_endpoints.getNearby(current_position, minimum_line_width);
+                nearest = nullptr;
+                coord_t nearest_dist2 = std::numeric_limits<coord_t>::max();
+                for(ExtrusionLine* candidate : nearby)
+                {
+                    if(processed_lines.find(candidate) != processed_lines.end())
+                    {
+                        continue; //Already processed this line before. It's linked to something else.
+                    }
+                    
+                    const coord_t dist2 = vSize2(candidate->junctions.front().p - current_position);
+                    if(dist2 < nearest_dist2)
+                    {
+                        nearest = candidate;
+                        nearest_dist2 = dist2;
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+
+    //The output walls from the skeletal trapezoidation have no known winding order, especially if they are joined together from polylines.
+    //They can be in any direction, clockwise or counter-clockwise, regardless of whether the shapes are positive or negative.
+    //To get a correct shape, we need to make the outside contour positive and any holes inside negative.
+    //This can be done by applying the even-odd rule to the shape. This rule is not sensitive to the winding order of the polygon.
+    //The even-odd rule would be incorrect if the polygon self-intersects, but that should never be generated by the skeletal trapezoidation.
```

Comment:
Checking polygon for self-intersections is naively an O(n²) algorithm. Not naively it can be done in O(n log n) time but this algorithm is [very complex](http://akira.ruc.dk/~keld/teaching/algoritmedesign_f03/Artikler/12/Chazelle92.pdf). Either way, you're dealing with loads of edge cases here that would make this quite complex endeavour.

It's not something I would like to implement in this ticket if it's just for an assertion. Note that Clipper's `offset()` function also doesn't deal with self-intersecting polygons so I think we'd have noticed by now if Arachne was outputting self-intersecting polygons.